packages/cosmic-swingset/lib/ag-solo/vats/lib-wallet.js

```diff
@@ -80,6 +79,43 @@ export async function makeWallet(
     inboxStateChangeHandler(getInboxState());
   }
 
+  // handle the update, which has already resolved to a record. If the offer is
+  // 'done', mark the offer 'complete', otherwise resubscribe to the notifier.
+  function updateOrResubscribe(id, offerHandle, update) {
+    const { updateHandle, done } = update;
+    if (done) {
+      // TODO do we still need these?
+      idToOfferHandle.delete(id);
+
+      const offer = idToOffer.get(id);
+      const completedOffer = {
+        ...offer,
+        status: 'complete',
+      };
+      idToOffer.set(id, completedOffer);
+      updateInboxState(id, completedOffer);
+      idToNotifierP.delete(id);
+    } else {
+      E(idToNotifierP.get(id))
+        .getUpdateSince(updateHandle)
+        .then(nextUpdate => updateOrResubscribe(id, offerHandle, nextUpdate));
+    }
+  }
+
+  // There's a new offer. Ask Zoe to notify us when the offer is complete.
+  async function subscribeToNotifier(id, offerHandle) {
+    E(zoe)
+      .getOfferNotifier(offerHandle)
+      .then(offerNotifierP => {
+        if (!idToNotifierP.has(id)) {
+          idToNotifierP.init(id, offerNotifierP);
+        }
+        E.when(E(offerNotifierP).getUpdateSince()).then(update =>
```

Comment:
is the result of notifier.getUpdateSince() always a promise? What is the Promise.resolve doing here? Just an extra assurance?

Reply:
The notifier is either a promise or a presence, so `E(offerNotifierP).getUpdatesSince()` returns a promise.  `E.when()` delivers `update` to `updateOrResubscribe()` after it resolves to a record.

I feel like I may have missed the point of your question, so this isn't what you wanted, please ask again.

Reply:
Is that what E.when() does? The comments on the code say it's equivalent to: `HandledPromise.resolve(x).then(res, rej)` and that it takes 3 arguments, so I think I don't understand. As I understand it, it's ensuring that x is a promise, whether or not x is originally, and then hooking up the res function as the callback for the promise resolving. 

If E(offerNotifierP).getUpdatesSince() returns a promise, can't we just do:

```
E(offerNotifierP).getUpdatesSince().then(update =>
```

Reply:
Yeah, it looks like I misread the doc. Going with your suggestion.